spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala

```diff
@@ -0,0 +1,199 @@
+        case updates if updates.isEmpty =>
+          col
+
+        // if there is an exact match, return the assigned expression
+        case Seq(update) if isExactMatch(update, col, resolver) =>
+          castIfNeeded(col, update.expr, resolver)
+
+        // if there are matches only for children
+        case updates if !hasExactMatch(updates, col, resolver) =>
+          col.dataType match {
+            case StructType(fields) =>
+              applyStructUpdates(col, prefixMatchedUpdates, fields, resolver, namePrefix)
+            case otherType =>
+              val colName = (namePrefix :+ col.name).mkString(".")
+              throw new AnalysisException(
+                "Updating nested fields is only supported for StructType " +
+                s"but $colName is of type $otherType"
+              )
+          }
+
+        // if there are conflicting updates, throw an exception
+        // there are two illegal scenarios:
+        // - multiple updates to the same column
+        // - updates to a top-level struct and its nested fields (e.g., a.b and a.b.c)
+        case updates if hasExactMatch(updates, col, resolver) =>
+          val conflictingCols = updates.map(u => (namePrefix ++ u.ref).mkString("."))
+          throw new AnalysisException(
+            "Updates are in conflict for these columns: " +
+            conflictingCols.distinct.mkString("[", ", ", "]"))
+      }
+    }
+  }
+
+  private def applyStructUpdates(
+      col: NamedExpression,
+      updates: Seq[ColumnUpdate],
+      fields: Seq[StructField],
+      resolver: Resolver,
+      namePrefix: Seq[String]): Expression = {
+
+    // build field expressions
+    val fieldExprs = fields.zipWithIndex.map { case (field, ordinal) =>
+      Alias(GetStructField(col, ordinal, Some(field.name)), field.name)()
+    }
+
+    // recursively apply this method on nested fields
+    val newUpdates = updates.map(u => u.copy(ref = u.ref.tail))
+    val updatedFieldExprs = applyUpdates(fieldExprs, newUpdates, resolver, namePrefix :+ col.name)
+
+    // construct a new struct with updated field expressions
+    toNamedStruct(fields, updatedFieldExprs)
+  }
+
+  private def toNamedStruct(fields: Seq[StructField], fieldExprs: Seq[Expression]): Expression = {
+    val namedStructExprs = fields.zip(fieldExprs).flatMap { case (field, expr) =>
+      Seq(Literal(field.name), expr)
+    }
+    CreateNamedStruct(namedStructExprs)
+  }
+
+  private def hasExactMatch(
+      updates: Seq[ColumnUpdate],
+      col: NamedExpression,
+      resolver: Resolver): Boolean = {
+
+    updates.exists(assignment => isExactMatch(assignment, col, resolver))
+  }
+
+  private def isExactMatch(
+      update: ColumnUpdate,
+      col: NamedExpression,
+      resolver: Resolver): Boolean = {
+
+    update.ref match {
+      case Seq(namePart) if resolver(namePart, col.name) => true
+      case _ => false
+    }
+  }
+
+  protected def castIfNeeded(
```

Comment:
I've reworked `castIfNeeded` completely. Now, it is closer to `checkField` in `TableOutputResolver`.

I should probably add this to the licence.

Reply:
Yes, please do.

Reply:
Done.